pkg/ssh/suftp.go

```diff
@@ -0,0 +1,145 @@
+/*
+Copyright 2023 ghostloda
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ssh
+
+import (
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/pkg/sftp"
+	"golang.org/x/crypto/ssh"
+)
+
+const promptpwd string = "gimmeyourpwdnow"
+const promptsuccess string = "thesudopwdwasok"
+const prompterror string = "ohmgsudofailed!"
+const promptlength int = 15
+
+func NewSudoSftpClient(conn *ssh.Client, sudopwd string, opts ...sftp.ClientOption) (*sftp.Client, error) {
+	s, err := conn.NewSession()
+	if err != nil {
+		return nil, err
+	}
+
+	// serverpaths are the most likely paths to the sftp-server binary, ordered from most likely less likely
+	// paths are from https://winscp.net/eng/docs/faq_su#fn2
+	serverpaths := []string{
+		"/usr/libexec/openssh/sftp-server",
+		"/usr/lib/openssh/sftp-server",
+		"/usr/lib/sftp-server",
+		"/usr/bin/sftp-server",
+		"/bin/sftp-server",
+		"sftp-server",
+	}
+
+	// sudo -p specifies the password prompt from sudo. The prompt is sent to stderr
+	// sudo -S dictates that password should be read from stdin
+	// sudo -u specifies the user
+	// sh -c 'cmd' passes cmd to sh
+	// >&2 echo "%s" echos promptsucess to stderr before starting sftp server ( see https://stackoverflow.com/a/23550347 )
+	// & %s is the final command to be executed is sudo is successful, which is the path to the sftp-server binary. It is a list of possible binary paths and it will pick the first one that exists. At the end, it will look in $PATH.
+	// || sh -c '>&2 echo "%s" & exit 1' -> runs if sudo failed and ensures that prompterror is written to stderr before erroring
+	// In total this means that we can read from stderr and write to stdin to
+	cmd := fmt.Sprintf(
```

Comment:
Using this approach may be enough https://github.com/sealerio/sealer/blob/main/utils/ssh/connect.go#L195

btw, do we really need to support ask-pass scenario?

Reply:
Most of the production environments do not grant sudo user NOPASSWD privileges.